aptos-move/block-executor/src/captured_reads.rs

```diff
@@ -287,6 +246,108 @@ impl<V: TransactionWrite> DataRead<V> {
     }
 }
 
+// Represents the result of comparing DataReads ('self' and 'other').
+#[derive(Debug)]
+enum DataReadComparison {
+    // Information in 'self' DataRead contains information about the kind of the
+    // 'other' DataRead, and is consistent with 'other'.
+    Contains,
+    // Information in 'self' DataRead contains information about the kind of the
+    // 'other' DataRead, but is inconsistent with 'other'.
+    Inconsistent,
+    // All information about the kind of 'other' is not contained in 'self' kind.
+    // For example, exists does not provide enough information about metadata.
+    Insufficient,
+}
+
+struct DataReadComparator {
+    // If set, BlockSTM V2 is enabled.
+    blockstm_v2_incarnation: Option<Incarnation>,
+}
+
+impl DataReadComparator {
+    fn new(blockstm_v2_incarnation: Option<Incarnation>) -> Self {
+        Self {
+            blockstm_v2_incarnation,
+        }
+    }
+
+    fn data_read_equals<V: PartialEq>(&self, v1: &DataRead<V>, v2: &DataRead<V>) -> bool {
+        if v1.get_kind() != v2.get_kind() {
+            return false;
+        }
+
+        match (v1, v2) {
+            (
+                DataRead::Versioned(v1_version, v1_value, v1_layout),
+                DataRead::Versioned(v2_version, v2_value, v2_layout),
+            ) => {
+                if v1_version == v2_version {
+                    true
+                } else {
+                    // TODO(BlockSTMv2): Like in MVDataMap, we assume data reads are not equal if both layouts
+                    // are set, in order to avoid expensive equality checks. This should be compensated here
+                    // by the above early return if versions are equal (for both V1 and V2 BlockSTM).
+                    self.blockstm_v2_incarnation.is_some()
+                        && v1_value == v2_value
+                        && v1_layout.is_none()
```

Comment:
Doesn't this cause data read validation failure on layouts being some? Can you point to where this is used for Block-STM v2 path?

Reply:
this is only used in post-commit materialization doing validation as a final safety check (the same flow we were using fetch_data instead of fetch_data_v2 even though in V2). if versions are equal we return true earlier, so we won't just fail validation for delayed field holder resources.

In data map, we always invalidate when a new write in inserted in between or removed and layouts are some, which should be rare for things containing delayed fields (and we should later improve the layouts anyway), and that also ensures in the end for such resources the versions should be the same.